runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java

```diff
@@ -222,6 +227,35 @@ DataflowPipelineJob run(Pipeline pipeline, DataflowRunner runner) {
     return Optional.<Boolean>absent();
   }
 
+  boolean checkMaxWatermark(DataflowPipelineJob job) throws IOException {
+    JobMetrics metrics = options.getDataflowClient().projects().jobs()
```

Comment:
Actually retry will happen in for loop outside of checkMaxWatermark. checkForSuccess() will check metrics first. If null, sleep for 10s and continue to next loop.